### ISSUE TITLE
Ideas for Route improvement

### ISSUE THREAD

**Issue by decebals** [decebals/pippo#254]
I want to move the discussion started by https://github.com/decebals/pippo/issues/253#issuecomment-185409221 here.

I will copy/paste the content of comment

> Yeah, I think that is good.
> 
> I also think we should consider doing deferred Route addition.
> 1. Observe the return type of any of [these methods](https://github.com/gitblit/fathom/blob/master/fathom-rest/src/main/java/fathom/rest/RoutesModule.java#L207).
> 2. Which is [this](https://github.com/gitblit/fathom/blob/master/fathom-rest/src/main/java/fathom/rest/RouteRegistration.java#L37) class.
> 3. And the asynchronous deferred Route [compilation/addition](https://github.com/gitblit/fathom/blob/master/fathom-rest/src/main/java/fathom/rest/RoutesModule.java#L315) step.
> 
> The basic idea is to allow a route to be configured declaratively before being added to the Router.  You seem to be heading that direction.

I want to say that from my review of RouteGroup implementation made by @ScienJus I observed that we must add the routes (DefaultRouter) in two steps:
1. `collect` all routes in Router (we can create a separate entity `RouteCollection` but for now Router is OK as home)

``` java
GET("/login", ...); // <-- application route

// OR
addRoute(...);

// OR
RouteGroup group = new RouteGroup("/useres");
group.GET("/{id}", ...); // <- group route
addRouteGroup(group);
```
1. `compile/process` added routes; in this point we know all details about collected routes (groups, ...) and we can do some preparation.

To implement two steps routes adding we need a way to call `Router.compileRoutes` method in the process of Application initialization.

``` java
 public final void init() {
        initializers.addAll(ServiceLocator.locateAll(Initializer.class));
        for (Initializer initializer : initializers) {
            log.debug("Initializing '{}'", initializer.getClass().getName());
            try {
                initializer.init(this);
            } catch (Exception e) {
                log.error("Failed to initialize '{}'", initializer.getClass().getName(), e);
            }
        }

        onInit();

        getRouter().compileRoutes(); // <-- HERE
    }
```

With this construction we can move `Route.setGroupUriPattern` in `compileRoutes`. Now it's safe, and also we can add some validation (see https://github.com/decebals/pippo/issues/248 for example) and take some decisions.

Now, my inline feedback for @gitblit :

> 1. Observe the return type of any of [these methods](https://github.com/gitblit/fathom/blob/master/fathom-rest/src/main/java/fathom/rest/RoutesModule.java#L207).
> 2. Which is [this](https://github.com/gitblit/fathom/blob/master/fathom-rest/src/main/java/fathom/rest/RouteRegistration.java#L37) class.

I prefer to not introduce a new entity `RouteRegistration` in project for this goal and to use `Route` for this. The `Route` class is for metadata. If we want a new entity, we can add `CompiledRoute` (it's a possible name) in this scenario we have `Route->Router.compileRoutes()->CompiledRoute`. We can discuss more about the content of `CompiledRoute` later if we agree that it's a good idea.

> 1. And the asynchronous deferred Route [compilation/addition](https://github.com/gitblit/fathom/blob/master/fathom-rest/src/main/java/fathom/rest/RoutesModule.java#L315) step.

Here, I see some improvements in the future :smile: My idea is to split the code and to add a functional `RouteCompiler`s (a possible name, or RouteProcessor). In this case in `Router` you can add programmatically (or via `ServiceLoader`)  a list with route compilers (`setRouteCompilers`) and on `Router.compileRouters` we iterate through this collection and call `compileRoute`. It's just an idea and I don't have the big picture yet.

**Comment by ScienJus** [comment]
it's great, `Route` and `RouteGroup` should be the tools to help people creating routes, then the real route will be only responsible store the metadata and used in url mapping.

**Comment by decebals** [comment]
I push some code (on a separate branch) related to this issue. The status is work in progress. I am a little busy with my new job but I wish to show you my early code because you can give me a feedback if I am on the right direction or not.
My next step is to try to use a `CompiledRouteTransformer`class  to implement metrics and route group (to validate the concept). After this, if you see value in this approach, I will try to improve the Controller using various `CompiledRouteTransformer` implementations.
I don't want to complicate the pippo core code without to put something (valuable) on table.    
I see a little drawback: you must call `compileRoutes()` on the router object in the unit tests after you add routes. Also, you are forced to add routes in `Application.init()`.
Any feedback is welcome.

**Comment by decebals** [comment]
I want to add more information about what I want to achieve with this modification.
I think that nowadays is good to be more declarative (see SpringBoot) and I like the result of [Fathom-Rest](http://fathom.gitblit.com/rest) (a wrapper around Pippo). In conclusion, I wish to transform `pippo-controller` module in something similar with `fathom-rest` and to preserve the `pippo-core` (for people that want to work with routes).
The actual `Route` class form Pippo is to keep/store the route's metadata. In `fathom-rest` the class that keep all route metadata si `RouteRegistration`.
The flow (take it from fathom-rest) is:
- gathering all controller classes
- process the controller's annotations 
- create the route metadata from above annotations and store these metadata in `RouteRegistation`
- process all `RouteRegistration`s (in `compileRoutes` method), transform some `RouteHandler`s (see routes with metrics), create and add the Pippo's `Route`s

We have this route correspondence (Pippo-Fathom):

Pippo | Fathom
-------- | --------------
Route | RouteRegistration
CompiledRoute | Route

You can see the code for `CompiledRoute` class [here](https://github.com/decebals/pippo/blob/compiled_route/pippo-core/src/main/java/ro/pippo/core/route/CompiledRoute.java).

My idea is to make `Route` more dynamic than `RouteRegistration`, to add the possibility for a user to add custom metadata (for example `metric`, or `secure`, ...). For this reason I added `bind(String key, Object value)` method  in Route class. Internally, the `Route` class store all custom metadata in `attributes: Map<String, Object>` and `bind(key, value)` is similar with `attributes.put(key, value)`.
You can add a custom `CompiledRouteTransformer` instance in `Router` if you want to process the `CompiledRoute` version for a specific `Route`.

The signature of `CompiledRouteTransformer` is very simple (a functional interface):
```java
/**
 * A {@code CompiledRouteTransformer} transform a compiled route.
 * For example you can modify the route name or the route handler.
 * If the returned route si null then the route will be removed from the compiled route list.
 *
 * @author Decebal Suiu
 */
public interface CompiledRouteTransformer {

    CompiledRoute transform(CompiledRoute compiledRoute);

}
```

The `Router` in `compileRoutes()` method creates an initial `CompiledRoute` instance for each `Route` and call `transform()` method for each registered `CompiledRouteTransformer`.

For example if I want to add __metric__ in a route I must add a custom metadata via `Route.bind(METRIC, ...)` and add a custom `CompiledRouteTransformer` (for example `MetricTransformer`) that decorate the initial `RouteHandler` with a `MetricRouteHandler`.

In Controller what we must to do is to scan for @Metric annotation and bind "metric" custom metadata to `Route` instance.

This is in few words what I try to achieve. Is it complicated? Is it achievable?

**Comment by decebals** [comment]
I updated branch [compiled_route](https://github.com/decebals/pippo/tree/compiled_route) with commit f362d818cb3736a4d09d30729e0b65ba82432b2c that show you the `CompiledRouteTransformer` (MettricsTransformer) in action. 
So, I can use the annotation approach (see MetricsTransformer) or the `Route.bind` approach (non annotation).  
I believe that Route -> Router.compileRoutes() -> CompiledRoute -> transform -> CompiledRoute is a powerful instrument that allow us to customize very easy, non intrusive, a RouteHandler.

Also, in the future, we can use `Route.bind(key, value)` to pass attributes/parameters from application to RouteHandler.

The next step is to try to implement the RouteGroup concept in a more clean fashion, using `CompiledRouteTransformer` and to finish #237 that depends by this issue.

@gitblit What you think? Do you see some problems or improvements?

**Comment by decebals** [comment]
I wish to finish the implementation of this feature. What you think? Do you see any drawback (speed, clarity, ...)? In your opinion is this feature valuable?
I think that the first benefice (is a simple one) of this feature is that it allows us to add a `RouteGroup` as a child for another `RouteGroup` and to resolve the routes from groups only at runtime (on compile routes). 
For example. take a look at [AdminRoutes](https://github.com/decebals/matilda/blob/master/src/main/java/ro/fortsoft/matilda/AdminRoutes.java), I don't like that I must to carry on the `Application` instance to be able to add a new RouteGroup.
I prefer something like:
```java
RouteGroup adminRoutes - new AdminRoutes();
adminRoutes.addGroup(new CustomerRoutes());
adminRoutes.addGroup(new CompanyRoutes());
adminRoutes.addGroup(new UserRoutes());
application.addRouteGroup(adminRoutes);
```

Right now, I use the below code (`AdminRoutes`):
```java
private void addCustomerRoutes() {
	// TODO (or a simple "addRouteGroup()")
	application.addRouteGroup(new CustomerRoutes());
}

private void addCompanyRoutes() {
    // TODO (or a simple "addRouteGroup()")
    application.addRouteGroup(new CompanyRoutes());
}

private void addUserRoutes() {
    // TODO (or a simple "addRouteGroup()")
    application.addRouteGroup(new UserRoutes());
}
```

The real benefice can be view in things like https://github.com/decebals/pippo/issues/237 (add super declarative mode to pippo-controller).

**Comment by decebals** [comment]
Resolved by #341.
